### PR TITLE
fetch snark work on start

### PR DESCRIPTION
**Pull request by svv232**
The design of this PR has changes from the original intention. This stems from the previous assumption that completed snark work is associated with a block. This is not the case, completed snark work, while added to a snarked ledger isn't directly associated to a block until it is included in one. This fact conflicts with a polling strategy that fetches all relevant snark work that is "close" to the best tip. Instead the approach is to fetch the most relevant work from available work selection work and return this instead during the RPC request.

### CODE REVIEW

**Comment by svv232** [comment]
!ci-build-me

**Comment by svv232** [comment]
!ci-build-me

**Comment by svv232** [comment]
!ci-build-me

File: src/lib/network_pool/snark_pool.ml [hunk 1]
```diff
@@ -561,6 +582,9 @@ end
 
 (* Only show stdout for failed inline tests. *)
 open Inline_test_quiet_logs
+module Mock_snark_pool =
```
Comment (mrmr1993) [inline src/lib/network_pool/snark_pool.ml:585]: This is generally a bad idea. Calling a functor also instantiates any 'toplevel code' in that module, and so any modules instantiated explicitly for tests should really be instantiated inside the relevant `test_module` (for `ppx_test`; equivalently in the test program for alcotest).

**Comment by svv232** [comment]
!ci-build-me

**Comment by svv232** [comment]
!ci-build-me

**Comment by svv232** [comment]
!ci-nightly-me

File: src/lib/mina_networking/mina_networking.mli [hunk 2]
```diff
@@ -109,6 +109,12 @@ module Rpcs : sig
       option
   end
 
+  module Get_completed_snarks : sig
+    type query = unit
+
+    type response = Transaction_snark_work.Checked.Stable.V2.t list option
```
Comment (deepthiskumar) [inline src/lib/mina_networking/mina_networking.mli:115]: Remove `Checked`? `Checked` modules are to signal that the value has been verified/checked, within a node. Across the network, nodes will/should verify responses upon receipt regardless

File: src/lib/network_pool/snark_pool.ml [hunk 3]
```diff
@@ -148,6 +151,15 @@ struct
                  ]
                :: acc ) )
 
+      let all_completed_checked_work t =
+        Map.fold ~init:[] !(t.snark_tables).all
```
Comment (deepthiskumar) [inline src/lib/network_pool/snark_pool.ml:155]: 1. You don't need to convert it to `Checked.t` (commented above). 
2. As discussed, this is a lot of snark work- all completed work referenced by the most recent 290 blocks in the best chain and any short forks within that range. In the max TPS case that'd be ~128*290*2 (assuming there are 2 blocks per slot, one canonical and one orphaned). Limiting helps but doesn't ensure that immediately relevant work is obtained. I think querying work referenced by recent, say 5 blocks in the best chain might be preferable.

**Comment by svv232** [comment]
!ci-build-me

File: src/lib/mina_networking/rpcs.ml [hunk 4]
```diff
@@ -744,6 +748,68 @@ module Get_transition_chain_proof = struct
...
+module Get_completed_snarks = struct
+  type nonrec ctx = ctx
+
+  module Master = struct
+    let name = "get_completed_snarks"
+
+    module T = struct
+      type query = unit [@@deriving sexp, to_yojson]
+
+      type response = Transaction_snark_work.Checked.Stable.V2.t list option
```
Comment (deepthiskumar) [inline src/lib/mina_networking/rpcs.ml:761]: This could be just a list, returning `[]` in the `None` case

**Review by deepthiskumar — APPROVED** [review]
Looks great!
I'm ok starting with first 32 work statements. Let's run a testnet and see how this fares

**Comment by svv232** [comment]
!ci-build-me

**Comment by svv232** [comment]
!ci-build-me

**Comment by svv232** [comment]
!ci-build-me

**Pull request merged**